src/graphql/graphql-api.php

```diff
@@ -0,0 +1,255 @@
+			}
+		}
+
+		return $flattened_blocks;
+	}
+
+	/**
+	 * Register types and fields graphql integration.
+	 *
+	 * @return void
+	 */
+	public static function register_types() {
+		/**
+		 * Filter to enable/disable the graphQL API. By default, it is enabled.
+		 * 
+		 * @param bool $is_graphql_to_be_enabled Whether the graphQL API should be enabled or not.
+		 */
+		$is_graphql_to_be_enabled = apply_filters( 'vip_block_data_api__is_graphql_enabled', true );
+
+		if ( ! $is_graphql_to_be_enabled ) {
+			return;
+		}
+
+		// Register the type corresponding to the attributes of each individual block.
+		register_graphql_object_type(
+			'BlockAttribute',
+			[
+				'description' => 'Block attribute',
+				'fields'      => [
+					'name'  => [
+						'type'        => [ 'non_null' => 'String' ],
+						'description' => 'Block data attribute name',
+					],
+					'value' => [
+						'type'        => [ 'non_null' => 'String' ],
+						'description' => 'Block data attribute value',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the individual block, with the above attribute.
+		register_graphql_type(
+			'BlockData',
+			[
+				'description' => 'Block data',
+				'fields'      => [
+					'id'          => [
+						'type'        => [ 'non_null' => 'ID' ],
+						'description' => 'ID of the block',
+					],
+					'parentId'    => [
+						'type'        => 'ID',
+						'description' => 'ID of the parent for this inner block, if it is an inner block. Otherwise, it will not be set. If it set, this will match the ID of the block',
+					],
+					'name'        => [
+						'type'        => [ 'non_null' => 'String' ],
+						'description' => 'Block name',
+					],
+					'attributes'  => [
+						'type'        => [
+							'list_of' => 'BlockAttribute',
+						],
+						'description' => 'Block attributes',
+					],
+					'innerBlocks' => [
+						'type'        => [ 'list_of' => 'BlockData' ],
+						'description' => 'Flattened list of inner blocks of this block',
+					],
+				],
+			],
+		);
+
+		// Register the type corresponding to the list of individual blocks, with each item being the above type.
+		register_graphql_type(
+			'BlocksData',
+			[
+				'description' => 'Data for all the blocks',
+				'fields'      => [
+					'blocks'   => [
+						'type'        => [ 'list_of' => 'BlockData' ],
+						'description' => 'List of blocks data',
+						'args'        => [
+							'flatten' => [
+								'type'        => 'Boolean',
+								'description' => 'Collate the inner blocks under each root block into a single list with a parent-child relationship. This is set to true by default, and setting it to false will reverse that to preserve the original block hierarchy, at the expense of knowing the exact depth when querying the inner blocks. Default: true',
+							],
+						],
+						'resolve'     => function ( $blocks, $args ) {
+							if ( ! isset( $args['flatten'] ) || true === $args['unflatten'] ) {
```

Comment:
bug, probably needs tests
```suggestion
							if ( ! isset( $args['flatten'] ) || false === $args['flatten'] ) {
```

Reply:
Good catch! I fixed this, tested it manually, and put it in https://github.com/Automattic/vip-block-data-api/pull/51/commits/d578fe470a0ae1248eb0198b682ef76d886bf3ff. I spent some time trying to add a test, but after my attempt I think @ingeniumed has already done a decent job trying to wrap the core logic from the resolver with a test:

https://github.com/Automattic/vip-block-data-api/blob/d578fe470a0ae1248eb0198b682ef76d886bf3ff/src/graphql/graphql-api.php#L220-L233

Our unit testing does not include the WPGraphQL plugin, so the logic that's in this resolver isn't directly tested [in `test_flatten_inner_blocks()`](https://github.com/Automattic/vip-block-data-api/blob/d578fe470a0ae1248eb0198b682ef76d886bf3ff/tests/graphql/test-graphql-api.php#L135). The bug here is in the thin shell of WPGraphQL code around the `flatten_inner_blocks()` function, which we test directly. Ideally we could have some end-to-end tests that include the request through WPGraphQL (like we do [for our REST tests](https://github.com/Automattic/vip-block-data-api/blob/d578fe470a0ae1248eb0198b682ef76d886bf3ff/tests/rest/test-rest-api.php#L130-L138)), but I think the testing here is a good enough start. As long as we catch the interface bugs, like this.